sympy/solvers/solveset.py

```diff
@@ -1355,10 +1532,13 @@ def add_type(eq, symbol, domain):
     if isinstance(rhs_s, FiniteSet):
         assert (len(rhs_s.args)) == 1
         rhs = rhs_s.args[0]
-        equation = Add(lhs, -rhs, evaluate=False)
 
         if lhs.is_Add:
-            result = add_type(equation, symbol, domain)
+            result = add_type(lhs, rhs, symbol, domain)
+        elif _is_lambert(lhs, symbol):
```

Comment:
`.is_Mul` and `.is_Pow` here would become specific to `lambert` types, therefore a direct identification of lambert type is done here